### PR TITLE
[BugFix] Vectorized priority update in replay buffers

### PR DESCRIPTION
**Pull request by matteobettini**
This is a patch for #1574

It fixes the core problem highlighted in that issue but there are still points that will need attention in a future refactoring of this class as I am not sure it is compatible with all the cases it aims to support. I am happy to elucidate more about this if needed.

### CODE REVIEW

File: torchrl/data/replay_buffers/replay_buffers.py [hunk 1]
```diff
@@ -682,6 +682,23 @@ def _get_priority(self, tensordict: TensorDictBase) -> Optional[torch.Tensor]:
...
         return priority
 
+    def _get_priority_vector(self, tensordict: TensorDictBase) -> torch.Tensor:
+        if "_data" in tensordict.keys():
+            tensordict = tensordict.get("_data")
+
+        priority = tensordict.get(self.priority_key, None)
+        if priority is None:
+            return torch.tensor(
+                [self._sampler.default_priority],
```
Comment (vmoens) [inline torchrl/data/replay_buffers/replay_buffers.py:692]:
```suggestion
                self._sampler.default_priority,
```

File: torchrl/data/replay_buffers/replay_buffers.py [hunk 2]
```diff
@@ -757,13 +774,16 @@ def update_tensordict_priority(self, data: TensorDictBase) -> None:
...
-                [self._get_priority(td) for td in data],
-                dtype=torch.float,
-                device=data.device,
-            )
+            if isinstance(data, LazyStackedTensorDict):
+                priority = torch.tensor(
+                    [self._get_priority_item(td) for td in data],
+                    dtype=torch.float,
+                    device=data.device,
+                )
```
Comment (vmoens) [inline torchrl/data/replay_buffers/replay_buffers.py:782]: I guess we assume that the stack dim is 0 but it could not be (?)
I think we can consider the priority can be stacked, no? At the end of the day it's supposed to be one priority per item. Maybe I'm missing something

File: torchrl/data/replay_buffers/replay_buffers.py [hunk 3]
```diff
@@ -1010,16 +1030,16 @@ def __call__(self, list_of_tds):
...
+        return tensor.max().item() if dim is None else tensor.max(dim=dim)[0]
     elif reduction == "min":
-        return tensor.min().item()
+        return tensor.min().item() if dim is None else tensor.min(dim=dim)[0]
     elif reduction == "mean":
-        return tensor.mean().item()
+        return tensor.mean().item() if dim is None else tensor.mean(dim=dim)
     elif reduction == "median":
-        return tensor.median().item()
+        return tensor.median().item() if dim is None else tensor.median(dim=dim)
```
Comment (vmoens) [inline torchrl/data/replay_buffers/replay_buffers.py:1042]: what about 
```python
    if reduction == "max":
        result = tensor.max(dim=dim)
    elif reduction == "min":
        result = tensor.min(dim=dim)
    elif reduction == "mean":
        result = tensor.mean(dim=dim)
    elif reduction == "median":
        result = tensor.median(dim=dim)
    else:
        raise NotImplementedError(f"Unknown reduction method {reduction}")
    if isinstance(result, tuple):
        result = result[0]
    return result
```

**Review by vmoens — APPROVED** [review]
LGTM, see my few minor comments

File: torchrl/data/replay_buffers/replay_buffers.py (hunk 3 above)
Comment (matteobettini) [reply torchrl/data/replay_buffers/replay_buffers.py:1042]: yes that works too, but we also need to call `.item()`

File: torchrl/data/replay_buffers/replay_buffers.py (hunk 2 above)
Comment (matteobettini) [reply torchrl/data/replay_buffers/replay_buffers.py:782]: this was the previous treatment so I am not super sure what was going on.

my guess is that it is assuming 0 as stack dim because in `expand` it stacks on 0.

are you suggesting to completely remove this and always go vectorized? I am down to try.

Comment (matteobettini) [reply torchrl/data/replay_buffers/replay_buffers.py:782]: maybe it was this way because the priority can have different shapes along the stack dim?

That is the only explanation I can guess

**Review by vmoens — APPROVED** [review]
LGTM thanks!
